src/server/announcements.ts

```diff
@@ -0,0 +1,107 @@
+import { nanoid } from 'nanoid'
+import { ActorInfo } from '../schemas'
+import ActivityPubSystem, { DEFAULT_PUBLIC_KEY_FIELD } from './apsystem'
+import { generateKeypair } from 'http-signed-fetch'
+import { APOrderedCollection } from 'activitypub-types'
+import { ActorStore } from './store/ActorStore'
+
+export class Announcements {
+  apsystem: ActivityPubSystem
+  publicURL: string
+
+  constructor (apsystem: ActivityPubSystem, publicURL: string) {
+    this.apsystem = apsystem
+    this.publicURL = publicURL
+  }
+
+  get actorUrl (): string {
+    return `${this.publicURL}/v1/${this.mention}/`
+  }
+
+  get outboxUrl (): string {
+    return `${this.actorUrl}outbox`
+  }
+
+  get mention (): string {
+    const url = new URL(this.publicURL)
+    return `@announcements@${url.hostname}`
+  }
+
+  getActor (): ActorStore {
+    return this.apsystem.store.forActor(this.mention)
+  }
+
+  async init (): Promise<void> {
+    const actorUrl = this.actorUrl
+    const actor = this.getActor()
+
+    try {
+      const prev = await actor.getInfo()
+      if (prev.actorUrl !== actorUrl) {
+        await actor.setInfo({
+          ...prev,
+          actorUrl
+        })
+      }
+    } catch {
+      const { privateKeyPem, publicKeyPem } = generateKeypair()
+      await actor.setInfo({
+        actorUrl,
+        publicKeyId: `${actorUrl}#${DEFAULT_PUBLIC_KEY_FIELD}`,
+        keypair: {
+          privateKeyPem,
+          publicKeyPem
+        },
+        announce: false
+      })
+    }
+  }
+
+  async announce (actor: string, info: ActorInfo): Promise<void> {
+    let existedAlready = false
+    try {
+      const existingActor = await this.apsystem.store.actorsDb.get(actor)
+      if (existingActor !== undefined) existedAlready = true
+    } catch (err) {
+      if (!(err as { notFound: boolean }).notFound) {
+        throw err
+      }
+    }
+
+    if (!existedAlready && info.announce) {
+      const activity = {
+        '@context': 'https://www.w3.org/ns/activitystreams',
+        type: 'Note',
```

Comment:
ActivityStreams hold activities, not Notes. This should be wrapped in a Create activity with the Note being the `object`.

e.g. here's how we link to notes in the staticpub example: https://github.com/RangerMauve/staticpub.mauve.moe/blob/default/outbox.jsonld#L13

We may need to store the notes in a separate store and add a new `/:actor/note/:id` URL to resolve them.

Reply:
I don't see a requirement in the [spec](https://www.w3.org/TR/activitystreams-core/) that states that OrderedCollections must hold Activitys. Actually, I was mostly copying off of Sutty's Jekyll implementation, which has plain Updates inside the outbox:
* https://sutty.nl/outbox.jsonld -> https://sutty.nl/lanzamiento-de-publicaciones-distribuidas-en-el-fediverso-a-trav%C3%A9s-de-sutty/update/1701726785.jsonld

Am I getting something wrong?

Reply:
Yeah, notice how the update you linked wraps the note in an Update activity. We should do the same here, but with a Create.

Reply:
We might, lets see if just wrapping is enough for now though.